Add LTS option to dam_break and parabolic_bowl test cases

gcapodag opened the pull request:
<!--
Thank you for your pull request.
Please add a description of what is accomplished in the PR here at the top:
-->
This PR enhances the `dam_break` and `parabolic_bowl` test cases with the option to run using the local time-stepping (LTS) time integrator. Note that LTS only works in a single layer setting. Please also see the companion PR https://github.com/E3SM-Project/E3SM/pull/6074
<!--
Below are a few things we ask you or your reviewers to kindly check. 
***Remove checks that are not relevant by deleting the line(s) below.***
-->
Checklist
* [ ] User's Guide has been updated
* [ ] Developer's Guide has been updated
* [ ] API documentation in the Developer's Guide (`api.rst`) has any new or modified class, method and/or functions listed
* [ ] Documentation has been [built locally](https://mpas-dev.github.io/compass/latest/developers_guide/building_docs.html) and changes look as expected
* [ ] The `E3SM-Project` submodule has been updated with relevant E3SM changes
* [ ] The `MALI-Dev` submodule has been updated with relevant MALI changes
* [ ] Document (in a comment titled `Testing` in this PR) any testing that was used to verify the changes
* [ ] New tests have been added to a test suite

<!--
Please note any issues this fixes using closing keywords: https://help.github.com/articles/closing-issues-using-keywords
-->

gcapodag commented:
This is `dam_break` with LTS, 120cm and default.

![dam_break_comparison](https://github.com/MPAS-Dev/compass/assets/10819458/70aef56d-d808-455f-ae87-029b131dbd48)

gcapodag commented:
This is `dam_break` with RK4, 120cm and default.
 
![dam_break_comparison](https://github.com/MPAS-Dev/compass/assets/10819458/6ffddce8-0f3d-4527-9985-908dff636379)

gcapodag commented:
While LTS seems to work better than RK4 with the coarser mesh (120 cm) it does not show the same improved behavior at finer meshes as the pictures below will show. It seems to be a thing though for the W/D scheme to not improve consistently with finer meshes as observed here: https://github.com/MPAS-Dev/compass/pull/670

gcapodag commented:
This is `dam_break` with LTS, 40cm and default.
 
![dam_break_comparison](https://github.com/MPAS-Dev/compass/assets/10819458/f5e77a9a-bed0-4879-8320-e02185a6f404)

gcapodag commented:
This is `dam_break` with RK4, 40cm and default.
![dam_break_comparison](https://github.com/MPAS-Dev/compass/assets/10819458/f8ee40fa-0a6b-4764-bf84-bb98be1cb52e)

gcapodag commented:
I am not showing the `ramp` case for `dam_break` and LTS because the results are qualitatively the same.

gcapodag commented:
An interesting thing for the `dam_break` test is that RK4 shows two wave fronts whereas LTS shows only one. I looked into the paper from which the `dam_break` test was taken: https://www.sciencedirect.com/science/article/abs/pii/S0098300413001362?via%3Dihub and it looks like after 2 seconds the wave has already left the domain between x in [0,4] and y in [0,2]. That also happens with LTS, but does not with RK4. See pictures below.

gcapodag commented:
Notice how the wave has moved away from the dark grey region in the picture from https://www.sciencedirect.com/science/article/abs/pii/S0098300413001362?via%3Dihub 
<img width="509" alt="Screen Shot 2023-11-17 at 11 21 59 AM" src="https://github.com/MPAS-Dev/compass/assets/10819458/a7e66c89-ac0b-41a2-aff0-20fc57b0c0b2">

This is LTS, and there is now wave front anymore in [10,13]x[11,13]:
<img width="235" alt="Screen Shot 2023-11-17 at 11 20 41 AM" src="https://github.com/MPAS-Dev/compass/assets/10819458/bb9b12e5-ab94-4151-8cc8-bd6d660d2d19">

A second wave front left seems to have barely just left the region [10,13]x[11,13] with RK4, I cannot tell for sure if the second wave front has left that region or not:
<img width="253" alt="Screen Shot 2023-11-17 at 11 20 28 AM" src="https://github.com/MPAS-Dev/compass/assets/10819458/ed9eaa9c-3789-4c29-b67a-eb40aaa78264">

gcapodag commented:
This is the solution at the final time for `parabolic_bowl` with LTS and ramp
 
![solution_360](https://github.com/MPAS-Dev/compass/assets/10819458/d886dad9-61fd-4c20-94bf-c9f184daa421)

gcapodag commented:
This is also LTS with `parabolic_bowl` and ramp

![points](https://github.com/MPAS-Dev/compass/assets/10819458/58e3ec5a-11f2-4159-8b92-74aeaa42070a)

gcapodag commented:
This is LTS with `parabolic_bowl` and ramp. The reference curve seems to be inverted with respect to the others but it should not be due to this PR and @sbrus89 is already looking into this

![error](https://github.com/MPAS-Dev/compass/assets/10819458/bc49797d-c032-43b8-b79e-d7f3469846d5)

cbegeman commented on compass/ocean/tests/dam_break/forward.py:
```diff
@@ -50,18 +54,37 @@ def __init__(self, test_case, resolution, name='forward', subdir=None,
-        self.add_input_file(filename='graph.info',
-                            target=f'{input_path}/culled_graph.info')
+        if use_lts:
+            self.add_namelist_options(
+                {'config_time_integrator': "'LTS'"})
+            self.add_namelist_options(
+                {'config_dt_scaling_LTS': "4"})
+            self.add_namelist_options(
+                {'config_number_of_time_levels': "4"})
```
You can combine this into a single dictionary argument with multiple entries

gcapodag replied:
thanks for your suggestion, can you point me to a case on compass where it is done?

gcapodag replied:
Done. Thanks!

gcapodag commented on compass/ocean/tests/dam_break/lts/lts_regions.py:
```diff
@@ -0,0 +1,328 @@
+        for icell in range(1, len(lines)):
+            if (lts_rgn[icell - 1] == 1 or lts_rgn[icell - 1] == 5):  # fine
+
+                # newf+= "0 1 0 " + lines[icell].strip() + "\n"
```
I don't think so, why?

cbegeman replied:
Oh, just that they can be deleted rather than commented out.

gcapodag replied:
Even though I do not anticipate using them in the future, I would rather have them there as a record in case we will need to use that old METIS partitioning for some reason. If I remove them then I will not remember how they were :)

cbegeman replied:
Ok. Why don't you leave a comment that mentions that they refer to the old METIS partitioning so others know why they are there?

cbegeman commented on compass/ocean/tests/parabolic_bowl/lts/lts_regions.py:
```diff
@@ -0,0 +1,332 @@
+                   logger=self.logger, use_progress_bar=use_progress_bar)
+
+
+def label_mesh(init, mesh, graph_info, num_interface,  # noqa: C901
```
Is this routine the same as used in `dam_break`? If so, I think it would be best to migrate it to the ocean framework.

gcapodag replied:
Yes except for the way the fine regions are setup. Once they are defined, everything else is the same. We are making a change to this routine with Jeremy to accommodate something that we need for Forward-Backward LTS, so once we do that we could also add an option to specify separately how to define the fine regions and add the rest of the routine to the ocean framework. For now, let's leave it like it is if you don't mind.

cbegeman replied:
Ok. As long as it is on your radar to get migrated to the framework when you're happy with it.

cbegeman commented on compass/ocean/tests/dam_break/forward.py:
```diff
@@ -50,18 +54,37 @@ def __init__(self, test_case, resolution, name='forward', subdir=None,
-        self.add_input_file(filename='graph.info',
-                            target=f'{input_path}/culled_graph.info')
+        if use_lts:
+            self.add_namelist_options(
+                {'config_time_integrator': "'LTS'"})
+            self.add_namelist_options(
+                {'config_dt_scaling_LTS': "4"})
+            self.add_namelist_options(
+                {'config_number_of_time_levels': "4"})
```
```suggestion
            self.add_namelist_options(
                {'config_time_integrator': "'LTS'",
                'config_dt_scaling_LTS': "4",
                'config_number_of_time_levels': "4"})
```
@gcapodag It would just look like this

gcapodag replied:
Thanks!

gcapodag commented on compass/ocean/tests/dam_break/lts/lts_regions.py:
```diff
@@ -0,0 +1,328 @@
+
+    max_area = max(area_cell)
+    min_area = min(area_cell)
+    max_width = 2 * np.sqrt(max_area / math.pi) / 1000
```
Done.

gcapodag commented on compass/ocean/tests/dam_break/namelist.forward:
```diff
@@ -13,4 +13,4 @@ config_drying_min_cell_height=1e-6
 config_vert_coord_movement='impermeable_interfaces'
 config_use_debugTracers=.false.
-config_disable_tr_all_tend = .true.
+config_disable_tr_all_tend =.true.
```
Done.

cbegeman replied:
FYI, when you have made the change, you can also go ahead and resolve the conversation.

gcapodag commented on compass/ocean/tests/parabolic_bowl/initial_state.py:
```diff
@@ -1,92 +0,0 @@
-from mpas_tools.io import write_netcdf
```
oops sorry, let me restore it

gcapodag commented:
@cbegeman thanks for your comments, sorry it took me a minute to address them. I just did. When you get a chance, please test this out using the companion LTS/WD PR in E3SM. Also, do you have any suggestions on how to fix the failures for CI? Thanks.

mark-petersen commented:
On the head of this PR, I try:
```
./conda/configure_compass_env.py --conda ~/miconda3 --compiler intel
source load_dev_compass_1.2.0-alpha.6_chrysalis_intel_openmpi.sh
compass list
```
and I get the error
```
    raise FileNotFoundError(f'Config file does not exist: {filename}')
FileNotFoundError: Config file does not exist: 
/gpfs/fs1/home/ac.mpetersen/repos/compass/pr/compass/ocean/tests/parabolic_bowl/parabolic_bowl.cfg
```
@gcapodag I suspect that the file 
```
compass/ocean/tests/parabolic_bowl/parabolic_bowl.cfg
```
exists locally for you but was not added to the repo. `git status` will show you all the uncommitted files. I've made that mistake too!

gcapodag commented:
Thanks @mark-petersen. I created this PR copying over files from another branch where there are extra files that should not be included here. It looks like that `parabolic_bowl.cfg` file did not make it with the other ones when I copied them. I added it now and pushed.

cbegeman commented:
@gcapodag I'm getting time-outs on my LTS parabolic bowl jobs with the automatically-generated job script (1 node, 2 hours) on chyrsalis (intel, impi) with the 5km-resolution forward step. Do you have a sense for whether we should increase the duration or node-count? We should set it to a safe value before merging this PR.

cbegeman commented:
@gcapodag All of the dam break tests ran fine for me on Chrys with intel, impi. I checked the viz and LTS looks visually identical to RK4 time integrator.

cbegeman commented:
> @gcapodag I'm getting time-outs on my LTS parabolic bowl jobs with the automatically-generated job script (1 node, 2 hours) on chyrsalis (intel, impi) with the 5km-resolution forward step. Do you have a sense for whether we should increase the duration or node-count? We should set it to a safe value before merging this PR.

Update: I got a timeout at 2h on 1 node for just the forward_5km step. It reached 0001-01-02_14:07:12.

gcapodag commented:
Thanks @cbegeman . Unfortunately I do not have access to Chrysalis and can only test on Perlmutter. On Perlmutter I ran the test `ocean/parabolic_bowl/standard/ramp_lts` first doing `salloc --nodes 1 --qos interactive --time 01:00:00 --constraint cpu` and then `compass run` and got this (the build was with DEBUG=FALSE):

>   test execution:      SUCCESS
>   test runtime:        31:53
> Test Runtimes:
> 31:53 PASS ocean_parabolic_bowl_standard_ramp_lts
> Total runtime 31:59
> PASS: All passed successfully!

I do not have a sense of what should be on Chrysalis since it looks like a 1 hr allocation is enough on Perlmutter. Considering that the run duration is 3 days and it only did 1 day and 14 minutes for the 5km case maybe we could start by requesting 2 nodes and if that is not enough also increase the time allocation. What do you think?

cbegeman commented:
@gcapodag I got a time out with debug on at `Doing timestep 0001-01-02_17:12:44` with 2 nodes and 2 hours of run time. This suggests to me that the code is hanging on chrysalis. It's a shame you don't have access to chrysalis. @mark-petersen Do you have thoughts on the best way to proceed?

gcapodag commented:
@cbegeman can you please try again on Chrysalis to see if it still hangs? I realized that I had not modified the forward namelist to set `ssh_gradient` as the pressure gradient type, which is the only one currently supported by LTS. Doing that, the plots at the different stations for the `dam break` test now come out the same as RK4.

cbegeman commented:
@gcapodag Great! That namelist fix did the trick for me on Chyrsalis. Approving now.

gcapodag commented:
Awesome! Thank you so much for working with me on this @cbegeman , you've been a huge help!

xylar commented:
@cbegeman, I'm assigning this to you to merge once it's been reviewed.  Does that work for you?

cbegeman commented:
Yes, happy to take care of the merge. @mark-petersen and @sbrus89 It appears that we are still waiting on your reviews.

sbrus89 reviewed (commented):
I ran these cases successfully on Perlmutter with debug flags and everything ran successfully and looks good! My only comment is that maybe the documentation for these cases should be updated to reflect the addition of LTS configs?

gcapodag commented:
Thank you @sbrus89 ! That is a good observation. I do not have a preference so I am going to defer to you and the others. One thing I would like to mention is that these tests serve as a check that W/D works with LTS and do not leverage any "LTS power" in terms of running faster, from this point of view, having LTS on is just a matter of using a different time integrator.

xylar commented:
@gcapodag, it seems worth creating an `lts.txt` suite in:
https://github.com/MPAS-Dev/compass/tree/main/compass/ocean/suites
and adding these test cases to it.  It might make sense to add the tests to the `wetdry.txt` suite as well but that's up to you, @cbegeman and @sbrus89 to decide.

I would still add something about LTS support to the docs for these tests wouldn't hurt.  You can say that it's just for regression testing and doesn't provide a performance improvement.

gcapodag commented:
Sounds good everyone. I just created an `lts.txt` suite and also added the lts tests to the `wetdry.txt` suite. I noticed that the parabolic bowl tests were not part of the `wetdry.txt` suite so I added those in as well, please let me know if not having them there was intentional and I will go ahead and remove them. I also updated the documentation with a little blurb on lts. Thanks!

sbrus89 reviewed (approved):
@gcapodag, this looks great to me. Nice job on getting LTS working with wetting/drying!

cbegeman commented:
> These are the errors for LTS with parabolic_bowl for both no ramp and ramp (compare to https://github.com/MPAS-Dev/compass/pull/670). The reference curve seems to be inverted with respect to the others but it should not be due to this PR and @sbrus89 is already looking into this

@sbrus89 @gcapodag Can one of you post an issue for this when you have a chance so we don't lose track of it? Thanks!

gcapodag commented:
@sbrus89 did you see that issue with the plot in your latest tests or is it only on my end?

sbrus89 commented:
@gcapodag, I saw it on my end too, but haven't had a chance to fix it. 

@cbegeman, I'll make an issue.

mark-petersen commented:
Tested by rebasing this pr on the head of main, then on chicoma
```
compass list | grep dam_break
 107: ocean/dam_break/40cm/default_lts
 108: ocean/dam_break/40cm/ramp_lts
 109: ocean/dam_break/40cm/default
 110: ocean/dam_break/40cm/ramp
 111: ocean/dam_break/120cm/default_lts
 112: ocean/dam_break/120cm/ramp_lts
 113: ocean/dam_break/120cm/default
 114: ocean/dam_break/120cm/ramp
compass setup -p ~/repos/E3SM/wettingdrying_lts_gcapodag/components/mpas-ocean -w $n/dambreak_wd_lts -n 111 112 113 114
```
and using the head of https://github.com/E3SM-Project/E3SM/pull/6074 compiled with gnu optimized. This proceeds through the steps, and I can see the same images as above. Beautiful!

![dam_break_comparison](https://github.com/MPAS-Dev/compass/assets/4036816/6e89ea20-c87b-43fc-9eed-c3dc49326736)

I then tested:
```
compass list | grep bowl
 368: ocean/parabolic_bowl/standard/ramp_lts
 369: ocean/parabolic_bowl/standard/ramp
 370: ocean/parabolic_bowl/standard/noramp_lts
 371: ocean/parabolic_bowl/standard/noramp
compass setup -p ~/repos/E3SM/wettingdrying_lts_gcapodag/components/mpas-ocean -w $n/bowl_lts -n 368 369 370 371
```
and it also runs fine. Thanks!

mark-petersen reviewed (approved):
@gcapodag thank you for the addition of these insightful test cases to show that wetting and drying work with LTS. This is a great example of coordinated code and testing PRs. The documentation `*.rst` pages also look great.

mark-petersen commented:
Also tested both `dam_break` and `parabolic_bowl` on chrysalis without any trouble.

gcapodag commented:
Great, thanks @mark-petersen !

cbegeman commented:
Merged! Thanks for all your work on this @gcapodag!

gcapodag commented:
Thanks @cbegeman ! I see that the E3SM PR received all the necessary approvals yesterday so it should be merged any day now I believe.
